uefi-raw/src/protocol/scsi.rs

```diff
@@ -150,7 +208,7 @@ pub struct ExtScsiPassThruProtocol {
     ) -> Status,
     pub reset_channel: unsafe extern "efiapi" fn(this: *mut Self) -> Status,
     pub reset_target_lun:
-        unsafe extern "efiapi" fn(this: *mut Self, target: *const u8, lun: u64) -> Status,
```

Comment:
This one also feels "logicially mut" to me, can we make this mut and ditto for the corresponding `uefi` wrapper?

Reply:
I can make the raw API mut, **but**...:

The entity in the safe `uefi` wrapper that corresponds to this funcitonality is `ScsiDevice<'a>` in function `ScsiDevice::reset()`.
And the only way to get an instance of `ScsiDevice<'a>` is through the device iterator: `ExtScsiPassThru::iter_devices()`.

How would you suggest to make this mut? Have a second `ExtScsiPassThru::iter_devices_mut()`?

Reply:
I think `ScsiDevice::execute_command` should also be `&mut`. So maybe there should only be `iter_devices_mut`, and drop `iter_devices` (since `iter_devices` indicates you will need to probe with `execute_command` anyway). Would that work, or do you see problems in API usage if we require exclusive access?

Reply:
Oh, and a general note: even if we decide to make the `uefi` interface `&` instead of `&mut`, we can still use `*mut` in the raw interface. Making a mut pointer from a non-mut reference is allowed, and doesn't affect soundness in and of itself.

Reply:
I just rewrote the API to use mutable and I ran into some ugly problems:

When you call `iter_devices_mut()` to produce `ScsiDevice`s, you mutably borrow the protocol (`ExtScsiPassThru`) itself for the lifetime of the generated `ScsiDevice`. Thus, while a device exists, you can't interact with the protocol itself. That produces all kinds of usability issues.

```rust
for mut device in scsi_pt.iter_devices_mut() {
            let request = ScsiRequestBuilder::read(scsi_pt.io_align())
                .with_timeout(Duration::from_millis(500))
                ...
```
![image](https://github.com/user-attachments/assets/81a91dc4-b817-4ae0-b8d8-9610ae5dce65)

Reply:
Thanks for experimenting and analyzing this. I don't have the capacity right not unfortunately to provide good guidance. @nicholasbishop  can you help, please?